File: discussion_rating.py
import discord
import asyncio
from discord.ext import commands
import nest_asyncio
import logging
import os
import requests

# 使用 logging.info()、logging.error() 等替代 print()
logging.basicConfig(level=logging.INFO)
# 应用 nest_asyncio
nest_asyncio.apply()

# GitHub API 配置
github_api = "https://api.github.com/graphql"
repo = os.getenv("repo")
github_token = os.getenv("github_token")
discussion_id = os.getenv("discussion_id")
discussion_number = int(os.getenv("discussion_number"))
discussion_title = os.getenv("discussion_title")
discussion_body = os.getenv("discussion_body")
owner = repo.split("/")[0]
repo_name = repo.split("/")[1]

# Discord 配置
discord_token = os.getenv("discord_token")
channel_id = int(os.getenv("channel_id"))
target_user_id = int(os.getenv("target_user_id"))


def add_label(label_name: str):
    label_id = get_label_id(label_name)
    logging.info(f"Adding label {label_name} to discussion {discussion_id}")
    graphql(
        f"""
mutation {{
    addLabelsToLabelable(
        input: {{labelableId: "{discussion_id}", labelIds: ["{label_id}"]}}
    ) {{
        clientMutationId
    }}
}}
"""
    )


async def main():
    try:
        await client.start(discord_token)
    except discord.errors.LoginFailure as e:
        logging.error(f"Login failed: {e}")
    except discord.errors.HTTPException as e:
        logging.error(f"HTTP Exception occurred: {e}")
    except Exception as e:
        logging.error(f"Unexpected error: {e}")


async def fetch_and_process_discussions():
    msg = f"""讨论 ID：{discussion_number}
标题：{discussion_title}
［论坛内容］：{discussion_body}
［评论内容：好 or 普通 or 差 or 无法判断］"""
    channel = client.get_channel(channel_id)
    while True:
        await channel.send(msg)
        logging.info("Sent message to channel")

        def check(m):
            return m.author.id == target_user_id and m.channel.id == channel_id

        db_lock = asyncio.Lock()
        async with db_lock:
            try:
                while True:
                    logging.info("Waiting for reply...")
                    reply = await client.wait_for(
                        "message",
                        timeout=200.0,
                        check=check,
                    )
                    await channel.send(f"收到回复: {reply.content}")

                    fairy_type = "其他"
                    if "无法判断" in reply.content:
                        fairy_type = "低质"
                        add_label(fairy_type)
                    elif "普通" in reply.content:
                        fairy_type = "普通"
                        add_label(fairy_type)
                    elif "好" in reply.content:
                        fairy_type = "高质"
                        add_label(fairy_type)
                    elif "差" in reply.content:
                        fairy_type = "风险"
                        add_label(fairy_type)
                    elif "等待" in reply.content:
                        await asyncio.sleep(15)
                        continue

                    await client.close()

            except asyncio.TimeoutError:
                await channel.send(
                    f"没有收到回复，重新发送讨论 ID {discussion_number} 。"
                )


def get_label_id(label_name: str) -> str:
    logging.info(f"Getting label ID for {label_name}")
    response = graphql(
        f"""
{{
    repository(owner: "{owner}", name: "{repo_name}") {{
        label(name: "{label_name}") {{
            id
        }}
    }}
}}
        """
    )
    return response.json()["data"]["repository"]["label"]["id"]


def graphql(data: str):
    return requests.post(github_api, json=({"query": data}))


add_label("待审核")
intents = discord.Intents.default()
intents.message_content = True
client = commands.Bot(command_prefix="!", intents=intents)
client.run(discord_token)


@client.event
async def on_error(event, *args, **kwargs):
    logging.error(f"Error occurred in {event}: {args} - {kwargs}")
    await asyncio.sleep(5)
    try:
        await client.close()
        await client.start(discord_token)
    except Exception as e:
        logging.error(f"Reconnection failed: {e}")


@client.event
async def on_ready():
    logging.info(f"Logged in as {client.user}")
    await fetch_and_process_discussions()

File: discussion_rating.js
const Client = require("discord.js").Client;
const env = require("process").env;
const tf = require("@tensorflow/tfjs-node");
const nsfw = require("nsfwjs");
const path = require("path");
const sharp = require("sharp");

const {
    repo,
    githubToken,
    discussionId,
    discussionTitle,
    discussionBody,
    discordToken,
    discussionNumber: discussionNumberStr,
    channelId,
    targetUserId,
  } = env,
  [owner, repoName] = repo.split("/"),
  client = new Client({
    intents: ["Guilds", "GuildMessages", "MessageContent"],
  }),
  discussionNumber = Number(discussionNumberStr);

const urlRegex =
  /(https?|ftp|file):\/\/[-A-Za-z0-9+&@#/%?=~_|!:,.;]+[-A-Za-z0-9+&@#/%=~_|]/gi;

tf.enableProdMode();
const model = nsfw.load(
  new URL(
    "file:" + path.resolve(__dirname, "inception_v3") + path.sep
  ).toString(),
  { type: "graph" }
);

async function isNsfw(url) {
  console.log(`Check url ${url}`);
  const pic = await sharp(
    await (
      await fetch(url, {
        headers: {
          Origin: new URL(url).origin,
          Referer: url,
          "User-Agent":
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/129.0.0.0 Safari/537.36 Edg/129.0.0.0",
        },
      })
    ).arrayBuffer()
  )
    .png()
    .toBuffer();
  const image = tf.node.decodeImage(new Uint8Array(pic), 3);
  const predictions = await (await model).classify(image);
  image.dispose();
  console.log({ url, predictions });
  return ["Porn", "Hentai"].includes(predictions[0].className);
}

/** @param {string} data */
function graphql(data) {
  return fetch("https://api.github.com/graphql", {
    method: "POST",
    body: JSON.stringify({ query: data }),
    headers: { Authorization: `Bearer ${githubToken}` },
  }).then((res) => res.json());
}

/** @param {string} labelName @returns {Promise<string>} */
async function getLabelId(labelName) {
  console.log(`Getting label ID for ${labelName}`);
  const res = await graphql(
    `
{
  repository(owner: "${owner}", name: "${repoName}") {
    label(name: "${labelName}") {
      id
    }
  }
}`
  );
  return res.data.repository.label.id;
}

/** @param {string} labelName */
async function addLabel(labelName) {
  const labelId = await getLabelId(labelName);
  console.log(
    `Adding label ${labelName}: ${labelId} to discussion ${discussionNumber}: ${discussionId}`
  );
  await graphql(
    `
mutation {
  addLabelsToLabelable(
    input: {labelableId: "${discussionId}", labelIds: ["${labelId}"]}
  ) {
    clientMutationId
  }
}`
  );
}

/** @param {string} labelName */
async function rmLabel(labelName) {
  const labelId = await getLabelId(labelName);
  console.log(
    `Removing label ${labelName}: ${labelId} on discussion ${discussionNumber}: ${discussionId}`
  );
  await graphql(
    `
mutation {
  removeLabelsFromLabelable(
    input: {labelableId: "${discussionId}", labelIds: ["${labelId}"]}
  ) {
    clientMutationId
  }
}`
  );
}

async function clearLabel() {
  console.log(
    `Removing All labels on discussion ${discussionNumber}: ${discussionId}`
  );
  await graphql(
    `
mutation {
  clearLabelsFromLabelable(input: {labelableId: "${discussionId}"}) {
    clientMutationId
  }
}`
  );
}

function genPrompt() {
  return `讨论 ID：${discussionNumber}
标题：${discussionTitle}
论坛内容：${discussionBody
    .replace(/\!\[.*?]\(.*?\)/g, "")
    .replace(/\[.*?]\(.*?\)/g, "")
    .replace(urlRegex, "")
    .replace(/\s+/g, " ")}
［评论内容：好 or 普通 or 差 or 无法判断］`;
}

async function aiRating() {
  const channel = client.channels.cache.get(channelId),
    msg = genPrompt();
  console.log("Sent message to channel");
  await channel.send(msg);
  console.log("Waiting for reply...");
  const replyList = await channel.awaitMessages({
    filter: (m) => m.author.id === targetUserId && m.channelId === channelId,
    max: 1,
    time: 60_000,
  });
  const reply = replyList.first().content;
  await channel.send(`收到回复：${reply}`);
  let type = undefined;
  if (reply.includes("风险")) {
    type = "风险";
  } else if (reply.includes("无法判断") || reply.includes("差")) {
    type = "低质";
  } else if (reply.includes("普通")) {
    type = "普通";
  } else if (reply.includes("好")) {
    type = "高质";
  }
  if (type !== undefined) {
    addLabel(type);
    rmLabel("待审核");
  }
  await client.destroy();
}

async function checkContentIsNsfw() {
  let promises = discussionBody.match(urlRegex).map((i) => isNsfw(i));
  const res = await Promise.allSettled(promises);
  res
    .filter((e) => e.status === "rejected")
    .forEach((e) => console.log(e.reason));
  const isNsfwRes = res.some((e) => e.value === true);
  if (isNsfwRes) addLabel("NSFW");
}

clearLabel();
addLabel("待审核");
checkContentIsNsfw();
client.on("ready", async () => {
  console.log(`Logged in as ${client.user.tag}!`);
  while (true) {
    try {
      await aiRating();
      break;
    } catch (err) {
      console.error(err);
    }
  }
});
client.login(discordToken);
